### PR TITLE
Add Ridge Regression to Machine Learning

### LINKED ISSUE #12108: Add Ridge Regression to Machine Learning
### Feature description

###Feature: Ridge Regression with Regularization

### **Description**: 
The program implements Ridge Regression, a type of linear regression that includes an L2 regularization term to prevent overfitting and improve generalization. It uses gradient descent to optimize the feature vector (weights) while adjusting for a regularization parameter (λλ). The program collects a dataset containing Average Damage per Round (ADR) and player ratings from a CSV file, processes the data, and fits a ridge regression model to find the best-fit line that predicts ADR based on player ratings.
### Key Components:

- **Data Collection:**

- The program fetches a dataset from a remote CSV file containing player ratings and their corresponding Average Damage per Round (ADR).

-  **Feature Vector Initialization:**

- Initializes a feature vector (θθ) to zero, representing the weights for the regression model.

- Gradient Descent Optimization:

- Implements steep gradient descent to update the feature vector based on the calculated gradients, ensuring convergence to the optimal weights.

- Regularization:

- Applies L2 regularization to the gradient updates to penalize larger weights, enhancing model robustness against overfitting.

- Error Calculation:

- Computes the sum of square errors to evaluate model performance and adjusts the feature vector iteratively.

- Mean Absolute Error Calculation:

- Adds a utility to compute mean absolute error between predicted and actual outcomes, providing insight into model accuracy.

-Result Output:

- Displays the resultant feature vector, which represents the optimized weights for predicting ADR based on player ratings.

###  Benefits:

- **Improved Generalization:** By incorporating regularization, the model can generalize better to unseen data
- **Flexibility:** Users can adjust the regularization parameter (λλ) to balance between fitting the training data and avoiding overfitting.

This feature enhances the program's capability to predict player performance metrics while providing a clear mechanism for controlling overfitting through regularization.

### PR DESCRIPTION
**Pull request by Harmanaya**
### Describe your change:

Fixes #12108
Fixes #14016

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Add or change doctests? -- Note: Please avoid changing both code and tests in a single pull request.
* [ ] Documentation change?

### Checklist:
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms include at least one URL that points to Wikipedia or another similar explanation.
* [x] If this pull request resolves one or more open issues then the description above includes the issue number(s) with a [closing keyword](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue): "Fixes #ISSUE-NUMBER".

### CODE REVIEW

**Comment by Harmanaya** [comment]
@cclauss Could you please review this PR when you get a chance?

**Comment by cclauss** [comment]
@priya-sundaram-dev, please review this pull request vs.
* #14016

**Comment by priya-sundaram-dev** [comment]
Reviewed this against #14016 as requested. Both implement a gradient-descent Ridge with the same public shape (`fit`/`predict`/`compute_cost`). Quick comparison and one blocking correctness issue on this PR:

**Correctness (this PR, #12246):** `predict` and `compute_cost` call `feature_scaling(features)` again, re-computing mean/std from *whatever data is passed in*. That means predictions on new/unseen data are scaled by that data's own statistics, not the training set's — so `theta` (fit on the training scaling) is applied to differently-scaled inputs and the results are inconsistent. The fix is to persist the scaling from `fit`:

```python
def fit(self, features, target):
    features_scaled, self._mean, self._std = self.feature_scaling(features)
    ...

def predict(self, features):
    if self.theta is None:
        raise ValueError("Model is not trained yet. Call `fit` first.")
    features_scaled = (features - self._mean) / self._std
    return features_scaled.dot(self.theta)
```

Minor: `mean_absolute_error` is annotated `-> float` but returns a NumPy scalar (wrap in `float(...)`); the `__main__` demo prepends an intercept column that then gets std-normalized to 0 and L2-penalized, so there's no true unregularized bias term — worth a comment or explicit handling.

**vs #14016:** that PR ships cleaner separate pytest tests, but it also adds `machine_learning/__init__.py` and a `machine_learning/tests/` package with a `conftest.py`. This repo is a collection of standalone, doctested scripts rather than an installed package, so introducing package `__init__.py` + a tests dir diverges from the established convention (`doctest`/`pytest --doctest-modules` in-file). 

**Recommendation:** the single-file, in-file-doctest approach here (#12246) fits repo conventions better — but it needs the training-scale fix above before merge. If we'd rather take #14016 for its tests, those tests should move to in-file doctests and drop the new `__init__.py`/`tests/` scaffolding. Either way we should close one to avoid the duplicate. Happy to re-review once the scaling fix lands.

**Comment by cclauss** [comment]
I will put __awaiting changes__ and __bug__ labels on both issues and we will see who fixes what.  Tiime is limited because we will close all __awaiting changes__ pull requests before the October 1st start of Hacktoberfest.

**Review by cclauss — APPROVED** [review]
